### ISSUE TITLE
New instance of assertion `(nd->witnessHeight == -1) || (nd->witnessHeight == indexHeight - 1)' failed

### ISSUE THREAD

**Issue by daira** [zcash/zcash#4680]
From https://github.com/zcash/zcash/issues/4301#issuecomment-670701619 :

> I just got this again on master, commit [205c7b5](https://github.com/zcash/zcash/commit/205c7b5844856ecf2ac52f8605b10f855c863a1a) :
> 
> ```
> Thank you for running a testnet zcashd v3.1.0 node!
> You're helping to strengthen the network and contributing to a social good :)
> 
> In order to ensure you are adequately protecting your privacy when using Zcash,
> please see <https://z.cash/support/security/>.
> 
>      Downloading blocks | 1019322 (1019626 headers) / ~1031720 (98%)
>                         | [                                                  ]
>            Next upgrade | Canopy at block height 1028500, in around 7 days
>             Connections | 5
>   Network solution rate | 12.000 Sol/s
> 
> Mining is paused while downloading blocks.
> 
> Since starting this node 2 minutes, 22 seconds ago:
> - You have validated 120 transactions!
> - You have completed 0 Equihash solver runs.
> 
> [Press Ctrl+C to exit] [Set 'showmetrics=0' to hide]
> zcashd: wallet/wallet.cpp:1094: void CopyPreviousWitnesses(NoteDataMap&, int, int64_t) [with NoteDataMap = std::map<SaplingOutPoint, SaplingNoteData>; int64_t = long int]: Assertion `(nd->witnessHeight == -1) || (nd->witnessHeight == indexHeight - 1)' failed.
> Aborted
> ```
> 
> This is not the result of bad chainstate from a build before #4573, because the same code was syncing fine just before this. (I had Ctrl-C'd, edited zcash.conf to enable mining, and then restarted.)
> 
> I have the datadir if anyone wants to investigate (ask me for the download link).

**Comment by LarryRuane** [comment]
To reproduce this problem, you must have shielded notes (funds) in your wallet. Reproducing the problem will _not_ cause loss of funds. You can get shielded funds using the [testnet faucet](https://faucet.testnet.z.cash/), or you can [mine to a shielded address](https://zcash.readthedocs.io/en/latest/rtd_pages/zcash_mining_guide.html#spending-mining-rewards) (It's easy to mine on testnet).

Applying this patch makes the problem reproduce more reliably, but may not be necessary:

```
--- a/src/validationinterface.cpp
+++ b/src/validationinterface.cpp
@@ -98,7 +98,7 @@ void ThreadNotifyWallets(CBlockIndex *pindexLastTip)
         // Run the notifier on an integer second in the steady clock.
         auto now = std::chrono::steady_clock::now().time_since_epoch();
         auto nextFire = std::chrono::duration_cast<std::chrono::seconds>(
-            now + std::chrono::seconds(1));
+            now + std::chrono::seconds(10));
         std::this_thread::sleep_until(
             std::chrono::time_point<std::chrono::steady_clock>(nextFire));
 
```

Start `zcashd` with a data directory that isn't fully synced with the blockchain. Watch the `debug.log` file (`tail -f` works well), and when you're seeing the usual stream of `UpdateTip: new best` lines, type control-c to stop `zcashd`. Then, restart. It will almost certainly hit the assertion failure within about one or two minutes.

Unfortunately, you need to run `zcashd -rescan` to repair your data directory (this process is faster on testnet).

**Comment by clayt0nk** [comment]
This has been happening to me for months(?), and I hoped it would just go away with a new version, but now I am up to v4.0.0 so I "tried something". Per things said in #4201 I tried a --reindex, which failed quickly. Per things said here I tried a --rescan, also failed quickly. My impression is if I toss wallet.dat, maybe that might work, BUT, will that not mean permanent loss of the 1.x shielded tokens in the wallet?

**Comment by LarryRuane** [comment]
I'm sorry this has happened. Please copy your `wallet.dat` and `zcash.conf` files out of your data directory, remove the data directory, create the directory (empty), restore those two files, then start zcashd. It will sync with the network (initial block download) and all should be well. If you follow these steps, you will not lose funds. 

Please advise here if this worked (or not), thank you.

**Comment by clayt0nk** [comment]
I will try. At the current rate of advance, it will require several days.

**Comment by LarryRuane** [comment]
@clayt0nk, have you had better luck?

**Comment by clayt0nk** [comment]
Well, progress, but after running overnight for three of the last four nights, zcashd is reporting only 20% completion, and zcash-cli is IIRC reporting only a bit more than 0.1. I think this might take weeks rather than days. On an entirely tangential note: zcashd used to stop almost immediately after a zcash-cli stop. Now it is taking a bitcoin-qt-esque several minutes to shut itself down.

**Comment by clayt0nk** [comment]
Survey says: rebuilding blockchain from scratch does not work. I have hit the same witnessHeight == indexHeight assertion around about block 861847 after deleting the blockchain and starting over. It is only a little bit more than one ZEC in the wallet. If some zcash dev who is working in this area would like me to send them the wallet.dat to wrestle with, that is fine with me, just ping me if anyone is interested. Especially if there is any possibility that there is still an iceberg in there waiting to eat someone else's tokens. I am not sure if the point was to signal software immaturity, but I am quite glad zcash did not encrypt the wallet, as that caused me to hold the vast majority of my ZEC elsewhere.

**Comment by LarryRuane** [comment]
@clayt0nk, thank you for your diligence in helping us fix this! Earlier you [wrote](https://github.com/zcash/zcash/issues/4680#issuecomment-695139563):
> Per things said here I tried a --rescan, also failed quickly.

In what way did it fail, the same assertion? At the same problem block height, 861847? By any chance, was the rescan interrupted (not run to completion in a single run)? 

We haven't seen a case where a `-rescan` didn't fix the problem. Would you mind trying `-rescan` again? If that still fails, then, yes, I guess the next step is to get a copy of your `wallet.dat` and I could try to reproduce it, and if so, dig in and analyze the problem. I can't think of another way. I promise you won't lose any funds.

**Comment by clayt0nk** [comment]
> We haven't seen a case where a -rescan didn't fix the problem
Happily, it seems I am not going to end that run, getblockchaininfo now shows verificationprogress = 0.99999..... :)
> In what way did it fail, the same assertion?
no, but since I did not open a fresh ticket, I do not seem to have preserved the exact error, but I am quite sure it was different
> was the rescan interrupted
I think what probably did in the -rescan on the old blockchain was that I first tried a -reindex, and whatever state that left things in when it failed, -rescan could not sort it out. (Which might be a bug in it's own right.....)

Thanks for being on top of this! Since '13 I have used many wallets and quite a collection of full nodes on many blockchains, and I have still, never once, lost money to a wallet fault. That too is a run that did not end today, and is a credit to the entire greater crypto community. :)
